ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala

```diff
@@ -0,0 +1,183 @@
+ */
+abstract class Parameter[T](val name: String) {
+  /**
+   * At construction, ensure that the given value is legal for this parameter.
+   */
+  assert(isValid, "Illegal value " + get() + " for strategy parameter " + companion.name)
+  
+  /** Retrieve the value for this parameter */
+  def get(): T
+  
+  /** The companion object of this parameter */
+  protected def companion: ParameterTrait[T]
+  
+  /**
+   * Whether or not this particle instance has been
+   * parameterized with a legal value.
+   */
+  def isValid: Boolean = companion.isValid(get())
+  
+  override def toString: String = get() + " (type: " + companion.name + ")"
+}
+
+
+
+/**
+ * "Static" methods that should be implemented in each concrete
+ * parameter's companion object.
+ * If the designer of a strategy wishes to change any of the fields
+ * in this trait, it is easy to extend it and override the field.
+ */
+trait ParameterTrait[T] {
+  /** Make a new instance of the associated parameter */
+  def getInstance(v: T): Parameter[T]
+  
+  /** Name of this parameter type */
+  def name: String = this.getClass.getName
+  
+  /**
+   * @return Whether the value is suitable for this parameter
+   */
+  def isValid(value: T): Boolean
+
+  /**
+   * Generate the range of valid values for this parameter.
+   * TODO: How to handle infinite ranges? Use streams?
+   */
+  def validValues: Iterable[T]
+  
+  /**
+   * Default value for this parameter
+   */
+  def defaultValue: T
+  
+  override def toString = name
+}
+
+
+
+/**
+ * Parameter representing a floating point coefficient in range [0; 1]
+ */
+case class CoefficientParameter(coefficient: Double) extends Parameter[Double]("Coefficient") {  
+  def companion = CoefficientParameter
+  def get(): Double = coefficient
+}
+
+object CoefficientParameter extends ParameterTrait[Double] {
+  def getInstance(v: Double) = new CoefficientParameter(v)
+  
+  /**
+   * Coefficient must lie in [0; 1]
+   */
+  def isValid(v: Double): Boolean = (v >= 0.0) && (v <= 1.0) 
+  
+  // TODO: handle user-selected resolution for these values
+  def validValues: Iterable[Double] = {
+    val resolution = 0.01
+    for {
+      n <- 0 to (1 / resolution).toInt
+    } yield (n * resolution)
+  }
+  
+  def defaultValue = 1.0
+}
+
+
+
+/**
+ * Parameter representing a pair of currencies to be traded.
+ */
+case class CurrencyPairParameter(currencies: (Currency, Currency)) extends Parameter[(Currency, Currency)]("CurrencyPair") {  
+	def companion = CurrencyPairParameter
+	def get(): (Currency, Currency) = currencies
+}
+
+object CurrencyPairParameter extends ParameterTrait[(Currency, Currency)] {
+	private type CurrencyPair = Tuple2[Currency, Currency]
+	def getInstance(v: CurrencyPair) = new CurrencyPairParameter(v)
+  
+	/**
+	 * All currency pairs are acceptable as long as they're not twice the same.
+	 * @TODO: It would be great if we could enforce an order inside the pair, so that we avoid having to handle swapped currency pairs
```

Comment:
Oh, so all strategies should be robust to a currency swap in their parameters. That makes sense.